ccm_web/client/src/utils/enrollmentValidators.ts

```diff
@@ -0,0 +1,113 @@
+import { ClientEnrollmentType, isValidRole } from '../models/canvas'
+import { InvalidationType } from '../models/models'
+import { DuplicateIdentifierInRowsValidator, RowInvalidation } from '../utils/rowValidation'
+import {
+  emailSchema, firstNameSchema, lastNameSchema, loginIDSchema, validateString, ValidationResult
+} from '../utils/validation'
+
+export interface EnrollmentInvalidation extends RowInvalidation {}
+
+interface EnrollmentRowsValidator {
+  validate: (values: string[]) => EnrollmentInvalidation[]
+}
+
+const getMessage = (result: ValidationResult, fieldName: string): string => {
+  return result.messages.length > 0 ? result.messages[0] : `Value for ${fieldName} is invalid.`
+}
+
+export class EmailRowsValidator implements EnrollmentRowsValidator {
+  validate (emails: string[]): EnrollmentInvalidation[] {
+    const invalidations: EnrollmentInvalidation[] = []
+    emails.forEach((email, i) => {
+      const emailValidationResult = validateString(email, emailSchema)
+      if (!emailValidationResult.isValid) {
+        invalidations.push({
+          rowNumber: i + 2,
```

Comment:
can you add a comment why `i+2`

Reply:
It's `i + 2` because `i` is an array index and JS uses zero-index arrays, so we add + 1. The column is not considered a data row, so we add + 1 again.

Reply:
Oh, you mean in the code, sure I can add a comment.

Reply:
yes

Reply:
Okay, as discussed via DM, I just extracted this calculation as `getRowNumber` so I didn't have to repeat this comment everywhere. https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/270/commits/8bbec4409aff223262e1d16898be8a7e84ff41a4

This has minimal utility, but may help us keep things consistent or improve readability. And I don't have to repeat the same comments.